File: java/src/main/java/sn/innolink/kyvshield/KyvShieldException.java
package sn.innolink.kyvshield;

/**
 * Exception thrown by the KyvShield SDK when an API call fails or the
 * response cannot be parsed.
 */
public class KyvShieldException extends RuntimeException {

    private final int statusCode;
    private final String responseBody;

    /**
     * Creates an exception with a human-readable message only (no HTTP context).
     *
     * @param message description of the error
     */
    public KyvShieldException(String message) {
        super(message);
        this.statusCode = 0;
        this.responseBody = null;
    }

    /**
     * Creates an exception wrapping another throwable.
     *
     * @param message description of the error
     * @param cause   underlying cause
     */
    public KyvShieldException(String message, Throwable cause) {
        super(message, cause);
        this.statusCode = 0;
        this.responseBody = null;
    }

    /**
     * Creates an exception with the HTTP status code and raw response body.
     *
     * @param message      description of the error
     * @param statusCode   HTTP status code returned by the server
     * @param responseBody raw response body as a string
     */
    public KyvShieldException(String message, int statusCode, String responseBody) {
        super(message);
        this.statusCode = statusCode;
        this.responseBody = responseBody;
    }

    /**
     * Returns the HTTP status code associated with this error, or {@code 0} if the
     * error did not originate from an HTTP response.
     */
    public int getStatusCode() {
        return statusCode;
    }

    /**
     * Returns the raw HTTP response body, or {@code null} if not available.
     */
    public String getResponseBody() {
        return responseBody;
    }

    @Override
    public String toString() {
        if (statusCode != 0) {
            return "KyvShieldException(statusCode=" + statusCode + "): " + getMessage();
        }
        return "KyvShieldException: " + getMessage();
    }
}

File: kotlin/src/main/kotlin/sn/innolink/kyvshield/KyvShieldException.kt
package sn.innolink.kyvshield

/**
 * Exception thrown by the KyvShield SDK.
 *
 * @param message Human-readable description of the error.
 * @param statusCode HTTP status code when the error originates from an HTTP response, or `null`.
 * @param errorCode Machine-readable error code returned by the API (e.g. `"MISSING_FIELD"`), or `null`.
 * @param responseBody Raw JSON response body as a string, if available.
 * @param cause Underlying exception that triggered this error, if any.
 */
class KyvShieldException(
    message: String,
    val statusCode: Int? = null,
    val errorCode: String? = null,
    val responseBody: String? = null,
    cause: Throwable? = null,
) : RuntimeException(buildMessage(message, statusCode, errorCode), cause) {

    companion object {
        private fun buildMessage(message: String, statusCode: Int?, errorCode: String?): String {
            val parts = mutableListOf(message)
            if (statusCode != null) parts.add("HTTP $statusCode")
            if (errorCode != null) parts.add("code=$errorCode")
            return parts.joinToString(" | ")
        }
    }

    override fun toString(): String =
        "KyvShieldException(message=${message}, statusCode=$statusCode, errorCode=$errorCode)"
}
